streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -196,6 +198,11 @@ boolean setState(final State newState) {
             }
 
             state = newState;
+            if (newState == State.PARTITIONS_ASSIGNED || newState == State.PARTITIONS_REVOKED) {
+                threadMetadata = new ThreadMetadata(threadMetadata.threadName(), state.name(), Collections.<TaskMetadata>emptySet(), Collections.<TaskMetadata>emptySet());
```

Comment:
Just for my own understanding: if state is PARTITION_ASSIGNED or PARTITION_REVOKED we assume that no new tasks go created yet and just we set both to empty set?

Should we do the same for PENDING_SHUTDOWN or DEAD? In case of failure, we might go from RUNNING to each of both without a PARTITION_REVOKED transition?

Reply:
Yes, tasks are just are either created or suspended when state is PARTITION_ASSIGNED or PARTITION_REVOKED. The ThreadMetadata should only expose currently running tasks.

I think you're right we sould do the same for non running states.